### PR TITLE
UCS/SIGNAL: signal and sigaction handlers were moved to separate module

### PR DESCRIPTION
**Pull request by hoopoepg**
- added separate module to handle signal and sigaction calls

### CODE REVIEW

**Comment by brminich** [comment]
```
OK ucs/config/ini.h
In file included from <stdin>:1:0:
ucs/debug/debug.h:11:22: fatal error: config.h: No such file or directory
 #  include "config.h"
                      ^
compilation terminated.
```

**Comment by hoopoepg** [comment]
yep, looking at this

**Review by brminich — APPROVED** [review]

File: configure.ac [hunk 1]
```diff
@@ -218,6 +218,7 @@ AS_IF([test "x$with_docs_only" = xyes],
      AM_CONDITIONAL([HAVE_AARCH64_HI1620], [false])
      AM_CONDITIONAL([HAVE_FUSE3], [false])
      AM_CONDITIONAL([IODEMO_CUDA], [false])
+     AM_CONDITIONAL([HAVE_SIGACTION_OVERRIDE], [false])
```
Comment (yosefe) [inline configure.ac:221]: let's remove this conditial
we added it initially to disable signal override in libucs.so, now it's a separate lib -- no longer needed

File: src/ucs/Makefile.am [hunk 2]
```diff
@@ -5,7 +5,7 @@
 # See file LICENSE for terms.
 #
 
-SUBDIRS = vfs/sock . vfs/fuse
+SUBDIRS = vfs/sock . vfs/fuse . signal
```
Comment (yosefe) [inline src/ucs/Makefile.am:8]: why added '.' another time?

File: src/ucs/signal/Makefile.am [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+# Copyright (C) 2022, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+#
+# See file LICENSE for terms.
+#
+
+if HAVE_SIGACTION_OVERRIDE
+
+lib_LTLIBRARIES = libucs_signal.la
+
+libucs_signal_la_CPPFLAGS = $(BASE_CPPFLAGS) $(BFD_CPPFLAGS)
```
Comment (yosefe) [inline src/ucs/signal/Makefile.am:11]: remove BFD

File: src/ucs/signal/Makefile.am [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+libucs_signal_la_CFLAGS   = $(BASE_CFLAGS) $(BFD_CFLAGS)
+libucs_signal_la_LDFLAGS  =  -version-info $(SOVERSION)
```
Comment (yosefe) [inline src/ucs/signal/Makefile.am:13]: extra space

File: src/ucs/signal/Makefile.am [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+libucs_signal_la_LIBADD   = $(top_builddir)/src/ucs/libucs.la
+
+libucs_signal_la_SOURCES = \
```
Comment (yosefe) [inline src/ucs/signal/Makefile.am:16]: no need newline

File: src/ucs/debug/debug.c [hunk 6]
```diff
@@ -1150,7 +1143,8 @@ sighandler_t signal(int signum, sighandler_t handler)
     return orig(signum, handler);
 }
 
-int sigaction(int signum, const struct sigaction *act, struct sigaction *oact)
+int ucs_debug_sigaction(int signum, const struct sigaction *act,
```
Comment (yosefe) [inline src/ucs/debug/debug.c:1146]: the logic of calling original handler should be part of libucs_signal.so
libucs_signal.so should call libucs function, to understand which parameters to pass to the original handler.
for example:
```c
int ucs_debug_is_signal_handler_forced(int signum) 
{
   return ucs_debug_initialized && ucs_debug_is_error_signal(signum);
}
```

and in libucs_signal.so, use this function result to determine behaior

File: src/uct/base/uct_iface.h [hunk 7]
```diff
@@ -7,6 +7,10 @@
 #ifndef UCT_IFACE_H_
 #define UCT_IFACE_H_
 
+#ifdef HAVE_CONFIG_H
```
Comment (yosefe) [inline src/uct/base/uct_iface.h:10]: why needed?

File: src/uct/tcp/tcp_net.c [hunk 8]
```diff
@@ -5,12 +5,12 @@
  * See file LICENSE for terms.
  */
 
-#include "tcp.h"
-
 #ifdef HAVE_CONFIG_H
 #  include "config.h"
 #endif
 
+#include "tcp.h"
```
Comment (yosefe) [inline src/uct/tcp/tcp_net.c:12]: unrelated?

File: configure.ac (hunk 1 above)
Comment (hoopoepg) [reply configure.ac]: > let's remove this conditial
> we added it initially to disable signal override in libucs.so, now it's a separate lib -- no longer needed

removed configuration

File: src/ucs/Makefile.am (hunk 2 above)
Comment (hoopoepg) [reply src/ucs/Makefile.am]: > why added '.' another time?

to be sure libucs_signal is built after libucs...
removed

File: src/ucs/debug/debug.c (hunk 6 above)
Comment (hoopoepg) [reply src/ucs/debug/debug.c:1159]: > the logic of calling original handler should be part of libucs_signal.so

ok, moved some evaluations into UCS_signal

File: src/ucs/signal/Makefile.am (hunk 3 above)
Comment (hoopoepg) [reply src/ucs/signal/Makefile.am]: > remove BFD

removed

File: src/ucs/signal/Makefile.am (hunk 4 above)
Comment (hoopoepg) [reply src/ucs/signal/Makefile.am]: > extra space

fixed

File: src/ucs/signal/Makefile.am (hunk 5 above)
Comment (hoopoepg) [reply src/ucs/signal/Makefile.am]: > no need newline

removed

File: src/uct/base/uct_iface.h (hunk 7 above)
Comment (hoopoepg) [reply src/uct/base/uct_iface.h]: > why needed?

below used macro ```IBV_HW_TM``` but config.h is not included. unrelated, removed for now, will create separate PR

File: src/uct/tcp/tcp_net.c (hunk 8 above)
Comment (hoopoepg) [reply src/uct/tcp/tcp_net.c]: > unrelated?

related, but this is workaround for issue. will apply another fix, more accurate

File: src/ucs/signal/Makefile.am [hunk 9]
```diff
@@ -4,16 +4,11 @@
...
-libucs_signal_la_CPPFLAGS = $(BASE_CPPFLAGS) $(BFD_CPPFLAGS)
-libucs_signal_la_CFLAGS   = $(BASE_CFLAGS) $(BFD_CFLAGS)
-libucs_signal_la_LDFLAGS  =  -version-info $(SOVERSION)
+libucs_signal_la_CPPFLAGS = $(BASE_CPPFLAGS)
+libucs_signal_la_CFLAGS   = $(BASE_CFLAGS)
+libucs_signal_la_LDFLAGS  = -version-info $(SOVERSION)
 libucs_signal_la_LIBADD   = $(top_builddir)/src/ucs/libucs.la
-
-libucs_signal_la_SOURCES = \
+libucs_signal_la_SOURCES  = \
```
Comment (yosefe) [inline src/ucs/signal/Makefile.am:13]: can be on one line

File: src/ucs/debug/debug.c [hunk 10]
```diff
@@ -1133,43 +1122,28 @@ static int ucs_debug_is_error_signal(int signum)
     return result;
 }
 
-sighandler_t signal(int signum, sighandler_t handler)
+sighandler_t ucs_debug_signal(int signum, sighandler_t handler)
```
Comment (yosefe) [inline src/ucs/debug/debug.c:1125]: can we fully move this function to libucs_signal.c, and not use it from ucs?
it should be called something like "ucs_orig_signal()"

File: src/ucs/signal/Makefile.am [hunk 11]
```diff
@@ -0,0 +1,14 @@
+#
+# Copyright (C) 2022, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+#
+# See file LICENSE for terms.
+#
+
+lib_LTLIBRARIES = libucs_signal.la
```
Comment (yosefe) [inline src/ucs/signal/Makefile.am:7]: can we have some test for it (that preloading libucs_signal.so prevents signals override)?
can also test manually and post output, or add CI test in next PR

File: src/ucs/sys/sys.h [hunk 12]
```diff
@@ -13,6 +13,10 @@
 #  include "config.h"
 #endif
 
+#ifndef __USE_GNU
```
Comment (yosefe) [inline src/ucs/sys/sys.h:16]: why needed here?

File: src/ucs/signal/configure.m4 [hunk 13]
```diff
@@ -1,8 +0,0 @@
-#
-# Copyright (C) 2022, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
-#
-# See file LICENSE for terms.
-#
-
-AC_CONFIG_FILES([src/ucs/signal/Makefile])
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucs/signal/configure.m4]: not needed if no customisation is used

File: src/ucs/debug/debug_int.h [hunk 14]
```diff
@@ -7,11 +7,29 @@
 #ifndef UCS_DEBUG_INT_H_
 #define UCS_DEBUG_INT_H_
 
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#ifndef __USE_GNU
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucs/debug/debug_int.h:14]: if we manage to move ucs_debug_get_orig_func to signal.c, than could remove it

File: src/ucs/sys/sys.h (hunk 12 above)
Comment (yosefe) [reply src/ucs/sys/sys.h:16]: maybe we could remove signal.h from sys.h, and use it only in the relevant C files (along with USE_GNU)

File: src/ucs/debug/debug_int.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/ucs/debug/debug_int.h]: > if we manage to move ucs_debug_get_orig_func to signal.c, than could remove it

removed

File: src/ucs/signal/Makefile.am (hunk 11 above)
Comment (hoopoepg) [reply src/ucs/signal/Makefile.am:7]: > can we have some test for it (that preloading libucs_signal.so prevents signals override)?
> can also test manually and post output, or add CI test in next PR

tested manually - signal/sigaction is hooked
will create test for it for CI

File: src/ucs/debug/debug.c [hunk 15]
```diff
@@ -1302,12 +1290,14 @@ static inline void ucs_debug_disable_signal_nolock(int signum)
     }
 
     original_action = kh_val(&ucs_signal_orig_action_map, hash_it);
-    ret = ucs_debug_sigaction(signum, original_action, &ucs_action);
+    /* remove signal from map prior to call sigaction to allow process it by
```
Comment (yosefe) [inline src/ucs/debug/debug.c:1293]: 👍 
minor: use upper case

File: src/ucs/signal/signal.c [hunk 16]
```diff
@@ -8,10 +8,35 @@
 #  include "config.h"
 #endif
 
+#ifndef __USE_GNU
+#  define __USE_GNU /* for sighandler_t */
+#endif
```
Comment (yosefe) [inline src/ucs/signal/signal.c:13]: now that it's move to C file, can we define __USE_GNU unconditionally - since no other file is included before it

File: src/ucs/signal/signal.c [hunk 17]
```diff
@@ -8,10 +8,35 @@
...
+#if !HAVE_SIGHANDLER_T
+#if HAVE___SIGHANDLER_T
+typedef __sighandler_t *sighandler_t;
+#else
+#error "Port me"
+#endif
+#endif
+
 
+static inline void *ucs_debug_get_orig_func(const char *symbol)
```
Comment (yosefe) [inline src/ucs/signal/signal.c:30]: does it have to be inline now?

File: src/ucs/debug/debug.c [hunk 18]
```diff
@@ -1252,9 +1188,11 @@ static void ucs_set_signal_handler(void (*handler)(int, siginfo_t*, void *))
     }
     sigemptyset(&sigact.sa_mask);
 
+    ucs_assert(!ucs_debug_initialized);
```
Comment (yosefe) [inline src/ucs/debug/debug.c:1191]: 👍 
maybe add comment?

File: src/ucs/debug/debug_int.h [hunk 19]
```diff
@@ -7,6 +7,10 @@
 #ifndef UCS_DEBUG_INT_H_
 #define UCS_DEBUG_INT_H_
 
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
```
Comment (yosefe) [inline src/ucs/debug/debug_int.h:12]: is it still needed?

**Comment by yosefe** [comment]
@bureddy FYI

File: src/ucs/debug/debug.c (hunk 18 above)
Comment (hoopoepg) [reply src/ucs/debug/debug.c:1193]: > 👍
> maybe add comment?

added comment

File: src/ucs/debug/debug.c (hunk 15 above)
Comment (hoopoepg) [reply src/ucs/debug/debug.c]: > 👍
> minor: use upper case

fixed

File: src/ucs/debug/debug_int.h (hunk 19 above)
Comment (hoopoepg) [reply src/ucs/debug/debug_int.h]: > is it still needed?

removed

File: src/ucs/signal/signal.c (hunk 16 above)
Comment (hoopoepg) [reply src/ucs/signal/signal.c]: > now that it's move to C file, can we define __USE_GNU unconditionally - since no other file is included before it

ok, removed

File: src/ucs/signal/signal.c (hunk 17 above)
Comment (hoopoepg) [reply src/ucs/signal/signal.c]: > does it have to be inline now?

removed ```inline```

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@hoopoepg pls squash
